### DISCUSSION TITLE
Websocket connection failed on community edition

### DISCUSSION THREAD

**Discussion by svandeneertwegh** [plausible/analytics#3826] [Self-Hosted Support] [answered]
### Past Issues Searched

- [X] I have searched open and closed issues to make sure that the bug has not yet been reported

### Issue is a Bug Report

- [X] This is a bug report and not a feature request, nor asking for self-hosted support

### Using official Plausible Cloud hosting or self-hosting?

Self-hosting

### Describe the bug

app-7a5f917ab82642c1…42aaea32.js?vsn=d:6 WebSocket connection to 'wss://<domain>/live/websocket?_csrf_token=X39ZCnwaKDAuBBQUETAuCSF…UBMB0e67tK-jA_ojlgIrKVd7BtfJKv&_mounts=0&_live_referer=undefined&vsn=2.0.0' failed: 

### Expected behavior

Succesful made websocket connections for small charts

### Screenshots

<img width="362" alt="Schermafbeelding 2024-02-25 om 11 06 40" src="https://github.com/plausible/analytics/assets/34129243/f0a51f7e-fc23-401f-9209-9fe96e19300e">

### Environment

```markdown
- OS: Ubuntu 23.10
- Browser: Google chrome
- Browser Version: latest
```

**Comment by ruslandoga** [comment]
👋 @svandeneertwegh 

Can you please share the latest logs from <kbd>docker compose logs plausible</kbd> and your networking configurations (reverse proxies, cdns, etc.)?

**Comment by svandeneertwegh** [comment]
plausible-1  | Loading plausible..
plausible-1  | Starting dependencies..
plausible-1  | Starting repos..
plausible-1  | create Plausible.Repo database if it doesn't exist
plausible-1  | create Plausible.IngestRepo database if it doesn't exist
plausible-1  | Creation of Db successful!
plausible-1  | Loading plausible..
plausible-1  | Starting dependencies..
plausible-1  | Starting repos..
plausible-1  | Running migrations for Elixir.Plausible.Repo
plausible-1  | Running migrations for Elixir.Plausible.IngestRepo
plausible-1  | Migrations successful!

**Comment by ruslandoga** [comment]
What about Google Chrome console logs? Does it include any kind of reason for the connection failure?

**Comment by svandeneertwegh** [comment]
![chrome-plausible-logs](https://github.com/plausible/analytics/assets/34129243/23dbab8d-f8d8-4417-9dcf-99581f970aa0)

**Comment by ruslandoga** [comment]
What about the "Console" tab?

**Comment by svandeneertwegh** [comment]
![chrome-plausible-logs2](https://github.com/plausible/analytics/assets/34129243/4a2840cc-c8f0-4ede-b299-6c2134a7f28f)

**Comment by ruslandoga** [comment]
Is there anything showed if you click on one of these log lines?

**Comment by svandeneertwegh** [comment]
No when clicking on it it attempts to open the wss link.. but doesnt work.

**Comment by ruslandoga** [comment]
I don't mean clicking on the link, but outside of it :) E.g. on the ➤ symbol.

**Comment by svandeneertwegh** [comment]
<img width="1595" alt="Schermafbeelding 2024-02-25 om 11 28 25" src="https://github.com/plausible/analytics/assets/34129243/a4846447-51fa-4ea7-9e37-a823f6882d1e">

**Comment by ruslandoga** [comment]
Do you use any proxies?

**Comment by svandeneertwegh** [comment]
I use nginx and use the tracker /script.js proxied. Maybe i need to add some websocket stuff to the nginx from the stats server?

**Comment by ruslandoga** [comment]
Most likely, yes :)

I'm going to move it to discussions since it doesn't seem to be an issue with Plausible itself.

**Reply by svandeneertwegh** [reply]
Great. Can you provide the websocket block in the nginx vhost for stats server?

**Reply by ruslandoga** [reply]
I'm not using Nginx myself, so I wouldn't know the exact way to do it. But maybe one of these would be useful to you:

- https://zaiste.net/network/nginx/howtos/nginx-config-elixir-phoenix/
- https://github.com/phoenixframework/phoenix/issues/663
- https://copyprogramming.com/howto/how-to-set-up-websockets-with-phoenix-behind-nginx#google_vignette

**Reply by svandeneertwegh** [reply, marked as answer]
I've solved it.

I added this:

proxy_http_version 1.1;
proxy_set_header Upgrade $http_upgrade;
proxy_set_header Connection "Upgrade";

To the location / block in nginx vhost where proxy_pass is set.

With the above setting active it does then support websockets.

**Reply by kurdin** [reply]
@svandeneertwegh Thank you!

**Reply by ruslandoga** [reply]
👋 @svandeneertwegh @kurdin 

Would you be able to review https://github.com/plausible/community-edition/pull/113?
Is that all that's required to make websockets work in nginx?

**Reply by kurdin** [reply]
@ruslandoga ~I don't see review button on the PR.~ Found it!
I created an update to the PR with adding more options to nginx reverse proxy settings
https://github.com/plausible/community-edition/pull/119

**Reply by shindi-renuo** [reply]
Amazing, thanks  ♥︎

**Comment by flipzoom** [comment]
### Plausible with Plesk + NGINX + WS (Websocket) Error
For all those who run the Plausible CE version on a server with Plesk, here is the way to get WS running in Plesk.

1. Go to `/var/www/vhosts/system/<yourdomainname>/conf/nginx.conf`
2. Find the block

```
#extension docker begin
location ~ ^/.* {
	proxy_pass http://0.0.0.0:8000;
	proxy_set_header Host $host;
	proxy_set_header X-Real-IP $remote_addr;
	proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
	proxy_set_header X-Forwarded-Proto $scheme;
}
#extension docker end
```
3. Copy this block
4. Go to `Your Domain > Websites & Domains > Dashboard > Docker Proxy Rules` and **remove** the rule created for Plausible

![Screenshot 2024-05-03 112601](https://github.com/plausible/analytics/assets/708183/08f641de-22ac-46ac-8178-3085f8796ab5)

5. Go to `Your Domain > Websites & Domains > Hosting DNS > Apache & nginx`
6. Deactivate the `proxy mode` and the `Serve static files directly by nginx`

![Screenshot 2024-05-03 113255](https://github.com/plausible/analytics/assets/708183/1ed5c386-b8b6-4d93-a567-f7d5aeb617dc)

7. Add the following and extended rule set under "Additional nginx directives"

```
location ~ ^/.* {
	proxy_pass http://0.0.0.0:8000;
	proxy_set_header Host $host;
	proxy_set_header X-Real-IP $remote_addr;
	proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
	proxy_set_header X-Forwarded-Proto $scheme;
	proxy_set_header Connection 'upgrade';
	proxy_set_header Upgrade $http_upgrade;
	proxy_http_version 1.1;
}
```
![Screenshot 2024-05-03 113520](https://github.com/plausible/analytics/assets/708183/94a2138f-b949-4395-89a8-2b9e202859b9)

8. Save your changes and restart nginx (via SSH `sudo systemctl restart nginx` or directly in Plesk `Tools & Settings > Services Management > Reverse Proxy Server (nginx)`)

The dashboard should now work with Websocket. When copying the Docker rules, make sure that the port is correct if you have specified a different one and that the IP address is 0.0.0.0. Do not make any changes to the vHost template directly, these will be overwritten by Plesk.

**Comment by ngist** [comment]
Just in case anyone else runs into this, if you have a location directive for `/` and `/live/websocket` as show in the [wiki](https://github.com/plausible/community-edition/wiki/Reverse-Proxy) it seems proxy_http_version 1.1; needs to be set for both locations. 

At least that's what it took on my setup, not entirely sure why though, perhaps it has to do with other flags I have set.
